### PR TITLE
website: A guide on serverless apps with Lambda and API Gateway

### PR DESCRIPTION
**Pull request by apparentlymart**
This is a common pattern that requires some non-obvious interactions between different resources, so we'll present it in tutorial form as a way to present all of the necessary parts together as a working system.

### CODE REVIEW

File: website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+EOF
+}
+```
+
+Each Lambda function must have an associated IAM role which dictates what
+access it has to other AWS services. The above configuration specifies a role
+with no access policy, effectively giving the function no access to any
+AWS services, since our example application requires no such access.
+
+Apply the above configuration as usual:
```
Comment (radeksimko) [inline website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: Shouldn't we also mention `terraform init` here?

File: website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+an earlier step:
+
+```hcl
+resource "aws_lambda_permission" "apigw" {
+  statement_id  = "AllowAPIGatewayInvoke"
+  action        = "lambda:InvokeFunction"
+  function_name = "${aws_lambda_function.example.arn}"
+  principal     = "apigateway.amazonaws.com"
+
+  source_arn = "arn:aws:execute-api:us-east-1:${data.aws_caller_identity.current.account_id}:${aws_api_gateway_rest_api.example.id}/*/*/*"
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: Yeah, the sequencing of the guide here was the reason I ended up not using that here. (Actually, I only remembered about `execution_arn` after I'd passed this point in the guide myself!)

I did consider restructuring the guide to do that step first but I wanted to complete the setup of the API before context-switching back to Lambda, because it feels confusing to keep switching back and forth between the two services.

File: website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md (hunk 1 above)
Comment (apparentlymart) [reply website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: A fair point... I assumed the user already knows how to do that, but there's not much harm in pointing it out again here to make sure.

**Comment by apparentlymart** [comment]
The point about the API gateway deployments is well-taken... I was aware of that limitation from my experiences implementing something like what's described here but had put it out of my mind because this pattern avoids dealing with it by keeping the API Gateway configuration very simple intentionally.

This guide as currently written is definitely walking a line between information and opinion, and the "just use API Gateway as a proxy and do all your routing logic inside the Lambda function" is a pretty strong opinion that I wasn't very explicit about. I'll revise a little to clarify that this is what's going on here, and weaken that sentence that makes a bold claim about how effective this is.

File: website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (apparentlymart) [reply website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: Actually now I re-read I remember that I already did re-work this slightly since my original draft and so the permission setup is already interleaved in there. It would actually make it _less_ context-switchy if I reordered it to do the deployment earlier.

That would then remove the need for the "test it in the API Gateway console" step, which is probably for the best since it's annoying to explain how to get there in the UI.

**Comment by apparentlymart** [comment]
I have re-ordered the steps so that the API Gateway REST API is now deployed before creating the Lambda permission. This removes the step of first testing within API Gateway, which does mean that this step does not have such a satisfying conclusion (can't see anything working yet) but avoids the need to explain how to traverse into the depths of the API Gateway console to find the test UI.

With that done, I was able to use the deployment's `execution_arn` rather than constructing an ARN from component parts.

If you'd be willing, I'd appreciate a re-read with the new structure in case I missed anything in the shuffle that makes this hard to follow.

**Comment by oscr** [comment]
@radeksimko @apparentlymart This is indeed a great addition and I would like to suggest two possible changes that might improve it. 

    Terraform is not a build tool, so the zip file must be prepared using a
    separate build process prior to deploying it with Terraform.

Yes, but Terraform can produce these zips. See [example](https://github.com/terraform-providers/terraform-provider-aws/blob/master/examples/lambda/main.tf) So **must** is not technically correct. 

The example also uses `aws_iam_policy_document` instead of using heredoc for the policy. Maybe that would also be a nice upgrade?

**Comment by apparentlymart** [comment]
Thanks for the feedback, @oscr!

Using `archive_zip` to create lambda source packages seems to be emerging as an anti-pattern because it prevents any reasonable rollback path. Part of the purpose of this guide is to establish as a best practice that application artifacts should always be built outside of Terraform, regardless of what target platform/scheduler is used. Similarly I would not recommend creation of AMIs with `aws_ami_copy` even though it is technically possible and useful in some rare cases. If I were to say anything about `archive_file` here I expect it would be to explicitly advise against using it! :grinning:

With guides like this it's always a trade-off to decide how many concepts to include. I intentionally skipped `aws_iam_policy_document` here since explaining it felt like a distraction from the main flow of the guide, while this way it should be immediately familiar to any reader who has worked with IAM policies.

Does that seem reasonable?

File: website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+~> **Warning:** Following this tutorial will create objects in your AWS account
+that will cost you money against your AWS bill.
+
+## Building the Lambda Function Package
+
+AWS Lambda expects a function's implementation to be provided as an archive
+containing the function source code and any other static files needed to
+execute the function.
+
+Terraform is not a build tool, so the zip file must be prepared using a
```
Comment (radeksimko) [inline website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: ❤️  💯  👍  TY

File: website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+steps then modify only the AWS Lambda function, leaving the API Gateway
+configuration unchanged.
+
+Due to API Gateway's staged deployment model, if you _do_ need to make changes
+to the API Gateway configuration you must explicitly request that it be
+re-deployed by "tainting" the deployment resource:
+
+```
+$ terraform taint aws_api_gateway_deployment.example
+```
```
Comment (radeksimko) [inline website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: Can we also mention that such redeployment will incur downtime here so folks are not surprised? I don't know what's the best wording, but this is mainly due to complex API and Terraform implementation which will (hopefully) change in the future so we can do more sane re-deployments of new versions with zero downtime.

Comment (apparentlymart) [reply website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md]: The downtime here is because the stage gets temporarily "un-deployed"? I didn't actually know that :grinning: so I guess I was just lucky that the downtime I've caused doing this occasionally in the past was not long enough for anyone to notice... or maybe having a CDN in front saved me.

In any case, yes... I will add a note about that here. I went to look at [the code](https://github.com/terraform-providers/terraform-provider-aws/blob/cefb0b0ce7f304853e6f3cb66880dbae1cc4cb2c/aws/resource_aws_api_gateway_deployment.go#L175-L188) to see what it's doing, and I think I'll just document plainly what's going to happen if you do this and thus leave it up to the reader to decide if they find that acceptable.

**Comment by oscr** [comment]
Hi there @apparentlymart! Thank you for that well written reply. 

While we could discuss if it's appropriate or not the point still stands: `the zip file must be prepared using a separate build process` is technically incorrect. A better choice of words would be something in the line of recommended/best practice. Agree?

Also I would like to argue that if it's an anti-pattern or not is a question of context. I don't disagree with you for the context that you describe. But in another context: such as delivering proof of concept to a customer it may be appropriate. What do you think?

In this case the example I think using `archive_zip` might be a good stepping stone towards the best practice you described before. Adding a warning / recommendation to migrate from it would suffice. Because lets be frank: using `archive_zip` is much faster and pleasant than doing it for hand. 

Agree with your argument for heredoc.
